File: frontend/src/app/model/DiscoverSongsDTO.ts
import { SongGenreDTO } from "./SongGenreDTO"
import { SongMoodDTO } from "./SongMoodDTO"

export interface DiscoverSongsDTO {
    pageable: Pageable,
    songs: SongDTO[]
}

export interface Pageable {
    pageNumber: number,
    pageSize: number,
    totalPages: number,
    totalElements: number,
    hasPrevious: boolean,
    hasNext: boolean,
    first: boolean,
    last:boolean
}

export interface SongDTO {
    encryptedId: string,
    title: string,
    authors: string[],
    genres: SongGenreDTO[],
    moods: SongMoodDTO[],
    thumbnailPath: string,
    releaseDate: string,
    soundLink: string
}

File: backend/dto/src/main/java/pl/dawid0604/mplayer/song/DiscoverSongsDTO.java
package pl.dawid0604.mplayer.song;

import java.util.List;

import static java.util.Collections.emptyList;

public record DiscoverSongsDTO(Pageable pageable, List<SongDTO> songs) {

    record SongDTO(String encryptedId, String title, List<String> authors,
                   List<SongGenreDTO> genres, List<SongMoodDTO> moods, String thumbnailPath,
                   String releaseDate, String soundLink) {

        public SongDTO() {
            this("", "", emptyList(), emptyList(), emptyList(), "", "", "");
        }
    }

    record Pageable(int pageNumber, int pageSize, int totalPages, int totalElements,
                    boolean hasPrevious, boolean hasNext, boolean first, boolean last) {

        public Pageable() {
            this(0, 0, 0, 0, false, false, false, false);
        }
    }
}
